### ISSUE TITLE
[Suggestion] PWM for Led driver dimming

### ISSUE THREAD

**Issue by leoneo** [kizniche/Mycodo#262]
Hello Kyle, 

I would like to control the brightness of 3 led strips (red, blue, white) in the time. 
For example function 1 : day 1 = 10%; day 2 = 15%; day 3 = 20 % ... day 15 to day 30 = 100%
function 2 : sunrise start at 6:30AM with 0% to 100% in 30min / sunset start at 8:30PM with 100% to 0% in 30min.

There are led drivers dimmable with 0-10Vc PWM input and with a 0-5V 0-10V PWM Converter Module it is theoretically possible with the Raspberry Pi to control the brightness. 
https://raspberrypi.stackexchange.com/questions/298/can-i-use-the-gpio-for-pulse-width-modulation-pwm
http://www.meanwell.com/productPdf.aspx?i=300
http://www.uctronics.com/0-5v-0-10v-to-0-100-voltage-to-pwm-converter-module.html

Please can you add a PWM output function in Mycodo ?

Thanks
Leo

**Comment by kizniche** [comment]
Good suggestion, How do you envision it being integrated into Mycodo? That is, what is your desired use? Such as x% at 8am, y% at 10am, etc.; response from sensors (conditional- if light sensor is x, set LED output to y, some sort of integration into the relay menu so relay actuation can somehow change brightness/color, or other?

**Comment by leoneo** [comment]
I use mycodo for the cultivation of microalgae. In the first stage the culture has a low density and requires little light. Excessive brightness inhibits growth. As cell density increases the luminous intensity must be increased. Then from day to day the luminous power increases. When cell density is at its maximum, microalgae are stressed with maximum intensity to make a red pigment.

I need a function that gradually increases the brightness during the first 15 days.

For an outdoor culture to have the ability to compensate for the loss of brightness by a cloud would be really very interesting to have a stable, climate-independent production.

I would like to use mycodo to a marine aquarium. For the coral it takes a function of sunrise and sunset every day.
(And the cherry on the cake would be the simulation of the lunar cycle.)

The PWM output function can be integrated in the relay menu.

**Comment by kizniche** [comment]
Here is what I currently think would work best for the 15 day cycle:
1. Create a method that gradually increases over the span of 15 days.
2. Modify the timers to accept a method as an input and a [special] relay as an output. This special relay will be a PWM relay.
3. When the timer is started, it will output according to the method.

For the sunrise/set, moon phase I could add a new type of controller that can calculate the sunrise/set and lunar phase for a given set of coordinates (and optional time offset to have a complete configuration), where there can be PWM output based on sunrise, midday, sunset, then switch to PWM for moon phase during night period. There could also be actions that can be taken at sunrise or sunset (or periods surrounding those using a time offset, such as "2 hours before sunrise" or "20 minutes after unset", activate relay ON, etc.)

**Comment by leoneo** [comment]
That will be great ;)

Do not forget the function of modifying the brightness of the LEDs according to the ambient brightness to maintain a constant luminosity in real time.

There is a lot of work.

**Comment by leoneo** [comment]
I know that the frequency of pulsed light can have a beneficial effect on photosynthesis. This frequency depends on the color (wavelength). I have to do some research on the subject.

**Comment by kizniche** [comment]
I was asking how is the hardware (LEDs) modulated by the software. And what software or code is used for that? I need a starting point as I don't have any hardware for LED PWM control.

**Comment by leoneo** [comment]
I think we can have the information in the datasheet of the driver. I am in the bed, I can find tomorrow.

**Comment by leoneo** [comment]
driver datasheet : Applying additive 10V PWM signal (frequency range 100Hz ~ 3KH)

you can test mycodo with 1 led and 1 resistor like that : 
https://circuitdigest.com/microcontroller-projects/raspberry-pi-pwm-tutorial

**Comment by wangine** [comment]
@kizniche Is the same thing i ask you last year if you can integrate a PWM output function in PID control . Can be 2 main forms of controll PID output (relays and PWM). Relays can do a good job for normal control ,but for fast changes below 1 second , relays just simply cant. I can handle with C but unfortunately my programming skills in Phyton and Flask are very low. (I'm hardware engineer) , software are just a passion. 
Here's a simple function for a PWM software (for processors without harware PWM module like Raspberry ) http://www.ccsinfo.com/forum/viewtopic.php?t=54585 .
Also the same function can be used in alternating power sources (~) not only on linear power sources (+-) like @leoneo  need .

**Comment by kizniche** [comment]
The RPI.GPIO module [has PWM](http://raspi.tv/2013/rpi-gpio-0-5-2a-now-has-software-pwm-how-to-use-it) and I think it would be somewhat easy to add it as a PID output (as the other use mentioned in this thread appears more difficult), now that I have a bit more experience w Python and the framework is more logical and accessible. What would be any additional PID options or changes needed to use PWM as an output? I'd assume I would essentially need to convert the PID output to a percentage, for use in defining the duty cycle. Any other suggestions would be appreciated, as I rarely work with PWM and don't currently have any devices that use it (though I can test the operation with my oscilloscope).

**Comment by wangine** [comment]
Yes Kyle , now i see, Pi is the same, software PWM (not hardware) and i assume can be selected as output any RPI pin. That would be great. Here is a tricky part. If controll is in one dirrection (raise or lower , usualy the temp) PWM output will be as percentage 0 - 100% and if the controll need to be in  both dirrection can be in two ways : 
1 : (the best one) with two RPI output pins , one for lower and one for raise . Raise between 0 - 100% and lower between 0 -100% . independently. Thats method is more easy to set the hardware without to much knowledge
2 : (need a complex hardware , specialy for alternating power source) . with one RPI pin . In that case raise will be between 50 - 100% and lower will be between 0 - 50% percentage.
So the PID interface will be without the relays radio buttons , also without durrations max and min.
We will need a 3 fields (radio buttons):
1: Field to select which method will use (relay or PWM)
2: Field to select the based PWM frequency (between 1Hz and 400kHz will be enoth for all aplications)
3: Field to specify the desired RPI pin (or pins in case of both dirrections).

**Comment by wangine** [comment]
@kizniche Also i can made a schematic for each one of the metods (2 pins or single RPI pin) and for the two power sources (alternating and countinuous ). You can fit into the help page after you finish.

**Comment by kizniche** [comment]
Sounds great. I can understand the 0-100% PID for a single relays (one 0-100% up, one 0- - 100 down), or just up or just down. However, I'm not fully understanding the other method you suggest. Schematics and descriptions would help. Thanks. I'm out right now, but plan to do some research when I get home later.

**Comment by wangine** [comment]
We can use the solid state relays instead of classic coil relay. That for heat sources load . Actualy is not to much difference between techniques of relays use. Classic relay can modulate in on and off (they can"t open or close only 30% or 66%) and solid state relays can work with PWM modulation . Thats mean if the power source wil be 120v AC , and relay is comanded with 50% duty cycle at 1kHz based frequency , the heat source will get 60v AC .
Some tips about how they work http://www.idc-online.com/technical_references/pdfs/electrical_engineering/The_Solid_State_Relay_Static_Relay_Overview.pdf
and here some relays i"ve used long time ago on my old regulation system with dsPIC and PWM output controller
http://www.farnell.com/datasheets/1754070.pdf?_ga=2.34697575.719620540.1500236779-1343759042.1500236779

**Comment by Ryobidelacroix** [comment]
What about a PWM IC? I found two examples. They can be controlled by either I2C or SPI.


https://www.adafruit.com/product/815
 - Code: https://github.com/adafruit/Adafruit-PWM-Servo-Driver-Library

https://www.adafruit.com/product/1429 
 - Code: https://github.com/adafruit/Adafruit_TLC5947

**Comment by wangine** [comment]
@Ryobidelacroix That's IC are specialised only for LED PWM , If the code will be made like @kizniche  suggestion http://raspi.tv/2013/rpi-gpio-0-5-2a-now-has-software-pwm-how-to-use-it the aplication will be general purpose , not only for LED. The native RPi pin"s can be used on any application.

**Comment by Ryobidelacroix** [comment]
Ah, even better!

**Comment by kizniche** [comment]
I completed about 75% of the PWM implementation last night, including the web interface and the PWM controller. Last is to test the operation and enable it as a new PID option. Then we can get into more complex uses.

**Comment by wangine** [comment]
I made a simple swap PWM code to simulate the output of PID on RPi . Schematics are for AC and DC load.
Unfortunately my PC cant handle  with record screen of 720p and ISIS simulate in real time, but is enough to someone understand. When you will finish i will test and made the schematic for any type on configuration (AC, DC , inductive , resistive load. ... )
https://www.youtube.com/watch?v=RPcaKuM9QoE

**Comment by kizniche** [comment]
I pushed the latest commit for what appears to be a working implementation of PWM. I say appears because I haven't tested it. I'll check with my oscilloscope tomorrow. There may very well be issues with this [PWM_support branch](https://github.com/kizniche/Mycodo/compare/PWM_support), as I haven't tested if current relay/PID functionality has been affected.

This commit only brings in the ability to add and configure a PWM output on the output's page (previously "relay" page), and turn on/off PWM. PID output will come next once I confirm there is indeed a PWM being output and normal relay/PID functionality hasn't been broken.

**Comment by wangine** [comment]
Ok. I put here some pictures with basical principle (the best and cheap way) . You are free to put in help page if you want.
Simulate the AC resistive load with PWM and 1% dyty cycle
![pwm 1 duty](https://user-images.githubusercontent.com/20494947/28337039-f5c1de80-6c0c-11e7-924f-8e1b52a31a0d.png)
Simulate the AC resistive load with PWM and 50% dyty cycle
![pwm 50 duty](https://user-images.githubusercontent.com/20494947/28337101-2a10935c-6c0d-11e7-8659-4a7e66cc9f98.png)
Simulate the AC resistive load with PWM and 99% dyty cycle
![pwm 99 duty](https://user-images.githubusercontent.com/20494947/28337119-342d6996-6c0d-11e7-93a7-2ecbf87d787a.png)
This is a basic principle , not a full point schematic, just to users to get the ideea how it works. Is the best way to controll the AC (most used) loads with PWM. It"s work with bipolar, MOSFET and IGBT transistors as well. Can controll any resistive load with less than 40% inductive .

**Comment by kizniche** [comment]
Great! I'll add them to the manual. I've never really used PWM besides simple uses with microcontrollers, so this is all relatively new to me.

**Comment by wangine** [comment]
Is just the principle , if you have a osciloscope is easy to fully understand how it's works and the hudge advantage than relays. With PID and PWM , dont stress the relays , the loads . Is almost linear . After you will finish the PWM code , i will upload the full real schematics (ISIS can't simulate the complex hardware, to much math for him), but only after i will test on my rPI . I use that for years on my old dsPIC MCU"s before discover your software.

**Comment by kizniche** [comment]
That last commit should have all parts of the system working with PWM, including PID output. I have my oscilloscope with me so I can test when I get home.

**Comment by wangine** [comment]
Ok. The first complete schematics with bipolar transistors, to controll with PWM and isolated output .
The normal circuit in phase with PWM , the most stable with bipolar transistor
![pwm 50 duty bipolar isolated no inverting phase](https://user-images.githubusercontent.com/20494947/28343701-47b7ef1c-6c27-11e7-993e-ae4a4c15769b.png)
And inverted phase than PWM , 
![pwm 50 duty bipolar isolated 1](https://user-images.githubusercontent.com/20494947/28343727-79fcf846-6c27-11e7-8391-a5602cfaa0c2.png)
Are tested only on resistive loads . Not on motors (fully inductive)

**Comment by kizniche** [comment]
Just tested turning on PWM through the output page (previously "relay" page). 

![newfile1](https://user-images.githubusercontent.com/838427/28345333-a1b6c43c-6bf7-11e7-86db-6b84c069069c.png)

![newfile2](https://user-images.githubusercontent.com/838427/28345337-a539e9c2-6bf7-11e7-8e2c-adfb387bf529.png)

**Comment by wangine** [comment]
That's perfect. I dont know which it is the limit of RPi , to generate PWM for everything to be fluid. To modulate AC sinusoidal, 50-60Hz , 100Hz isn't enough. Even for LED's we need more than 600Hz , to dont be visible. For AC is good 1kHz , but if is possible can be increased to avoid ressonance noise. If you can test on 22kHz and if you get that perfect square , that will be great and suitable for all aplications. Also need to be possible to put (like a variable) the desired frequvency and the PID will controll only the duty cycle 0 - 100. The PID dont need to controll the frequvency , that will be chosed by the user (depend of application) .That why i work on that schematic , to avoid RPi to deal with "zero cross reference" . Will be less work for you.

**Comment by kizniche** [comment]
I currently have the frequency set on the outputs configuration and, yes the PID only controls the duty cycle. I'm testing PID right now. I'll see about 22kHz and report back.

**Comment by kizniche** [comment]
I tested with 1kHz, 5kH, and 20kHz, below. Doesn't look great.

1 kHz

![newfile4](https://user-images.githubusercontent.com/838427/28347739-dfd8e754-6c06-11e7-99ec-ce03320eb540.png)

5 kHz

![newfile6](https://user-images.githubusercontent.com/838427/28347719-cdb0871c-6c06-11e7-9e2d-1e49dda07348.png)

20 kHz

![newfile7](https://user-images.githubusercontent.com/838427/28347724-d061040a-6c06-11e7-8ef4-9cdb42b63f7b.png)

**Comment by kizniche** [comment]
Strange, because I've found multiple [sources](http://codeandlife.com/2012/07/03/benchmarking-raspberry-pi-gpio-speed/) that state GPIO should be able to output up to 70 kHz.

**Comment by kizniche** [comment]
This script gives me a clean 200 kHz square wave on my scope:

```python
#!/usr/bin/python
# -*- coding: utf-8 -*-
import RPi.GPIO as GPIO

GPIO.setmode(GPIO.BCM)
GPIO.setup(23, GPIO.OUT)

while True:
    GPIO.output(23, True)
    GPIO.output(23, False)
```

**Comment by kizniche** [comment]
Yet this, like the code currently being used with Mycodo, when 20000 Hz is set, my scope only detects ~ 5 kHz.

```python
#!/usr/bin/python
# -*- coding: utf-8 -*-
import time
import RPi.GPIO as GPIO

GPIO.setmode(GPIO.BCM)
GPIO.setup(23, GPIO.OUT)

pwm = GPIO.PWM(23, 20000)
pwm.start(50)
time.sleep(60)
pwm.stop()
GPIO.cleanup()
```

**Comment by kizniche** [comment]
Using the hardware PWM pin 18, and pigpio, I was able to get a nice looking 200 kHz wave, and even was able to go up to into MHz.

```python
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Use hardware PWM pin 18 (BCM numbering)
import pigpio

frequency = 250000
duty_cycle = 50

pi = pigpio.pi()
pi.hardware_PWM(18, frequency, duty_cycle * 10000)
```

**Comment by kizniche** [comment]
I was able to get up to 40 kHz on a regular (non-hardware PWM pin), by restarting pigpiod with a sample rate of 1 μs (```sudo pigpiod -s 1```) and running this code.

```python
#!/usr/bin/python
# -*- coding: utf-8 -*-
import pigpio
import time

pin = 23  # Pin using BCM numbering
frequency = 40000
duty_cycle = 50

pi = pigpio.pi()

pi.set_PWM_frequency(pin, frequency)
calc_frequency = int((duty_cycle / 100.0) * 256)
pi.set_PWM_dutycycle(pin, calc_frequency) # 50%
time.sleep(60)
pi.stop()
```

see [here](http://abyz.co.uk/rpi/pigpio/python.html#set_PWM_frequency) for a table of available frequencies based on the selected sample rate when starting pigpiod. At 1 μs, available frequencies are: 40000 20000 10000 8000 5000 4000 2500 2000 1600 1250  1000   800  500  400  250  200  100   50

**Comment by kizniche** [comment]
So, I could provide the option of using either the hardware PWN pins (which would use the pigpio ```hardware_PWM()``` function), or a regular pin (which would use pigpio's ```set_PWM_frequency()``` and ```set_PWM_dutycycle()```).

There are a couple ways to do this. This could be in the form of a dropdown with two options, 1) hardware PWM pin and 2) regular pin. Then, if hardware pin i selected, the user could be presented with a list of available frequencies. If regular pin is selected, then the user can use any frequency they wish, up to 40 kHz (if the sample rate of pigpiod is changed to 1, which can be made default in the next release).

Anyway, that's enough testing for tonight. Everything seems to be working fine, we just need to sort out what PWM code to use and how to configure it in the web UI.

**Comment by wangine** [comment]
Yea , i know python cant do a good job vs C , but is enough 20kHz. And look ok below 20kHz , anyway no one can't hear the armonics at 16kHz modulation. Also the digital scopes missunderstand the triggers level sometimes, can not fully trust, personaly i use mixed analog+dig scope. Why you say doesnt look great because i dont see the altered square . Just a little trigger lack on the end. I assume you use that lib.
http://abyz.co.uk/rpi/pigpio/python.html#set_PWM_range
http://abyz.co.uk/rpi/pigpio/cif.html

**Comment by wangine** [comment]
Yea looks great . For some reason , your last post didn't show when i write.

**Comment by leoneo** [comment]
Good new, thanks you very much. 
Now I need to find hardware and test it.

**Comment by kizniche** [comment]
I said "doesn't look great" because with RPi.GPIO, setting 1 kHz yielded an 869 Hz wave, 5 kHz yielded a 2.92 kHz wave, and 20 kHz yielded a 5.4 kHz wave.

I'll be changing the code to use pigpio, since, in my tests at least, it's proven to be more stable, accurate, and produce a higher frequency.

**Comment by wangine** [comment]
Yea , i did't look at freqvency , only at square form and the level voltage of RPi pin

**Comment by wangine** [comment]
Sorry , maybe is not right place. How to update manual? I delete old system, i clone again , system remain the same . How to update that commit? 
Anyway i need to resolve my issue, daemon still stopping after several hours (that happen for days). I try to change sensors, RPi interface, power source , Even with new image and fresh Mycodo same issue. Tomorow i will reflash with the old stable version  (working continuosly from last year without any isssue on other RPi) and i will see.

**Comment by kizniche** [comment]
Okay, I think everything related to PWM is working fine now. The only thing left to do is work on the graphs (which I have not touched and are probably broken). Here is what I have:

The output page allows a new selection dropdown "PWM output" to be selected.

The PWM output has the following options: Library ("Any Pin, <= 40 kHz" and "Hardware Pin, <= 30 MHz"), BCM Pin, Frequency (Hertz), Current Draw, all which can be set. Then there is a "Turn on" button and a Duty Cycle field to turn on the PWM from the Output Page, sort of how the relays can be turned on for a duration of seconds.

![screenshot-192 168 0 5-2017-07-19-20-35-39](https://user-images.githubusercontent.com/838427/28395427-e7fda9b8-6cc1-11e7-90fd-729dde56e12b.png)

On the PID page, there is now a drop down selection for the PID type, which includes "Relay Output" and "PWM Output". The PWM PID includes the standard options like Input device, setpoint, PID gains, etc. But then only have "Output ID (Raise)" and "Output ID (Lower)" in place of the relay-specific options.

![screenshot-192 168 0 5-2017-07-19-20-26-56](https://user-images.githubusercontent.com/838427/28395270-c1c73094-6cc0-11e7-830a-bbb437cb1921.png)

The specific code that converts the control variable to a duty cycle is this:

```python
                        if self.control_variable > self.period:
                            self.raise_duty_cycle = 100.0
                        else:
                            self.raise_duty_cycle = float(
                                (self.control_variable / self.period) * 100)
```

I'm not sure if this is the best way to do this, but it was the first thing that came to mind and It seems to work fairly well. I'm open to other suggestions as to how to calculate the duty cycle. We can also have a dropdown to select different methods.

Let me hear any thoughts if there should be any changes.

> i clone again , system remain the same

This is because the code I'm pushing is on the [PWM_support branch](https://github.com/kizniche/Mycodo/tree/PWM_support), so you'll have to checkout that branch before installing normally. I didn't want to push these edits to master yet, before we iron out all the issues.

```
git clone https://github.com/kizniche/Mycodo
cd Mycodo
git checkout PWM_support
cd install
sudo /bin/bash ./setup.sh
```

Then, if you want to update to the latest commit (if I push more), do this:

```
sudo service mycodo stop
sudo /etc/init.d/apache2 stop
cd Mycodo
git pull
sudo /bin/bash ./mycodo/scripts/upgrade_commands.sh initialize
sudo /bin/bash ./mycodo/scripts/upgrade_commands.sh restart-web-ui
sudo /bin/bash ./mycodo/scripts/upgrade_commands.sh restart-daemon
```

When I finally make the next version release to master, you can just to go the upgrade page and it will upgrade your install to the latest release.

> daemon still stopping after several hours (that happen for days)

Does this happen on a freshly-installed system without anything added yet? Any errors in te daemon or HTTP logs?

**Comment by wangine** [comment]
Yes , thanks, i found the way."If you find yourself you never forget " //
Yes that happen in last 3 weeks, i just assume is my router fault when i try to acces over home network , Linksys keep rejecting my RPi  , i didn't take care because my tobacco wasn't ready to ferment, after i assume has something wrong with sensors. I change, RPi system (fresh lite and full also), new fresh mycodo install, i activate just one sensor , time by time and wait , i change sensors from other working RPi with very old and stable mycodo , same issue. I change power supply also. Dont worry , i will test tomorow with the old mycodo to see if is repeated. Now i wish to try PWM , is almost ready, i doit manualy with WinSCP . Testing..........

**Comment by kizniche** [comment]
I'll be constantly pushing commits to the branch (about to push one to fix a bug not turning the PWM signal off when the PID controller is stopped), so I'd recomment using the git clone method to be able to get the latest code.

**Comment by wangine** [comment]
Yes yes , now will be easy way , my network is faster . Daemon show randoms errors 
2017-07-16 13:31:51,154 - mycodo - INFO - 42.192 MB ram in use
2017-07-16 13:31:51,411 - mycodo.influxdb - ERROR - No measurement available in the past 60 seconds.
2017-07-16 13:31:51,415 - mycodo.pid_1 - WARNING - No data returned from influxdb
2017-07-16 13:32:22,801 - mycodo.sensors.sht1x_7x - ERROR - SHT1x7xSensor raised an exception when taking a reading: (113L, 255)
2017-07-16 13:36:22,517 - mycodo.sensors.sht1x_7x - ERROR - SHT1x7xSensor raised an exception when taking a reading: (214L, 255)
2017-07-16 14:10:45,435 - mycodo - INFO - Received command to terminate daemon

and other one saved a short part 

  File "/var/www/mycodo/env/local/lib/python2.7/site-packages/sqlalchemy/engine/strategies.py", line 97, in connect
  File "/var/www/mycodo/env/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 385, in connect
    return self.dbapi.connect(*cargs, **cparams)
OperationalError: (sqlite3.OperationalError) unable to open database file

2017-07-19 06:29:55,706 - mycodo.sensors.sht1x_7x - ERROR - SHT1x7xSensor raised an exception when taking a reading: (<functools.partial object at 0x640b3120>, 1.0)
2017-07-19 06:29:55,707 - mycodo.sensor_1 - ERROR - StopIteration raised. Possibly could not read sensor. Ensure it's connected properly and detected.
2017-07-19 06:30:10,786 - mycodo.sensors.sht1x_7x - ERROR - SHT1x7xSensor raised an exception when taking a reading: (<functools.partial object at 0x6608a390>, 1.0)

**Comment by kizniche** [comment]
Did you just copy your database over from another install?

**Comment by wangine** [comment]
No no . was a fresh one. Errors are from two RPi . One with lite and one with full image. On lite the install wasn't succesful , so i delete Mycodo in home and i reinstall again with

rm -r Mycodo
git clone https://github.com/kizniche/Mycodo
cd Mycodo/install
sudo /bin/bash ./setup.sh

**Comment by kizniche** [comment]
Oh, I thought you were pasting logs with errors from the PWM_support branch install.

I don't have the SHT1x7x sensor, so I can't really debug the issue myself. The "unable to open database" error is strange.

**Comment by kizniche** [comment]
I would be careful about cloning master. It can at any time be in a non-working state. I would always only download releases using the install instructions. The releases are the only thing I really strive to have a working system with.

**Comment by wangine** [comment]
You have a clue ? I think i messed up the system . 
![pwm error](https://user-images.githubusercontent.com/20494947/28396427-09660c02-6d04-11e7-86f5-96613e9c8176.png)

**Comment by kizniche** [comment]
Well, that's not a very descriptive error, lol

Do this... in Mycodo/mycodo/controller_relay.py, add this line after line 686, so the code looks like this:

```python
        except Exception as except_msg:
            self.logger.exception(1)
            return 1, "Add_Mod_Relay Error: ID {id}: {err}".format(
                id=relay_id, err=except_msg)
```

Then restart the daemon, ```sudo service mycodo restart``` and do what you did to get the error, again.

**Comment by kizniche** [comment]
There should then be the full exception traceback in the daemon log.

**Comment by wangine** [comment]
yea
`2017-07-20 04:41:55,475 - mycodo.relay - ERROR - 1
Traceback (most recent call last):
  File "/var/www/mycodo/mycodo/controller_relay.py", line 654, in add_mod_relay
    if self.relay_state(relay_id) != 'off':
  File "/var/www/mycodo/mycodo/controller_relay.py", line 845, in relay_state
    if self.pwm_state[relay_id]:
KeyError: 1
2017-07-20 04:45:48,072 - mycodo.relay - ERROR - 1
Traceback (most recent call last):
  File "/var/www/mycodo/mycodo/controller_relay.py", line 654, in add_mod_relay
    if self.relay_state(relay_id) != 'off':
  File "/var/www/mycodo/mycodo/controller_relay.py", line 845, in relay_state
    if self.pwm_state[relay_id]:
KeyError: 1`

**Comment by wangine** [comment]
I wonder , you can acces my RPi ? 
adress : wangine.go.ro
user    : test
pass    : 123456
I just forward my rpi IP on 80 port to my DDNs

**Comment by kizniche** [comment]
Nope. SSL will be port 443, also.

There aren't any errors when the daemon starts? ANd I assume those errors occur when you try to save the configuration.

**Comment by wangine** [comment]
yes same error. I difficut for me to focus, is 4.50 PM here, i just wait you because i know you are active arown that hour. Should enable 443 forward on my router ? I can doit in 10 seconds , UDP ?

**Comment by kizniche** [comment]
I found the issue. Making a commit to fix in a moment.

**Comment by kizniche** [comment]
Oh, you meant AM, not PM. Yes, that's too early for me, for a weekday. That latest commit should fix the issue.

**Comment by wangine** [comment]
working now . You can login in my RPi . I just forward 443. I test on my phone out of network

**Comment by kizniche** [comment]
I changed the "test" user password for security, so no one else can use it.

**Comment by kizniche** [comment]
Use the update instructions in [this comment](https://github.com/kizniche/Mycodo/issues/262#issuecomment-316564431) and see if that fixes it.

**Comment by wangine** [comment]
security dont matter , RPi is temporary for testing. Power line are OFF, so only sensors are active , 
OK, I give a try now

**Comment by wangine** [comment]
OK.will work . The user remain active for you , you can test sensors , i add the last one sht25. I change to 30sec read data. maybe 15sec wasn't good for them. I will test tomorow is altmost morning  (after i will sleep 4 hours ) on my service and i made a test board for AC and DC  . You do a great job for moment . Interface are cool and easy to configure.. I wish you good day or a night

**Comment by kizniche** [comment]
I think perhaps adding a Minimum and Maximum allowed duty cycle to be set for PWM PID controllers could be beneficial. I'm getting weird results at duty cycles of 0 and 100.

**Comment by kizniche** [comment]
Nevermind, it was a bug in the calculation of the duty cycle. All is well without those extra options.

**Comment by wangine** [comment]
Yes , will be . I can made more tests on other PI next hours on my service about PWM. My home RPi stoped with strange error after 40 minutes . I just seen daemon stoped when i wake up. User is active , You can figure out what's going on. If i move the sensors on my old RPi will run without error.

**Comment by kizniche** [comment]
I changed the database schema in one of the latest commits, so after you update, you'll need to delete ```/Mycodo/databases/mycodo.db``` and then visit the web interface to generate a new database with the correct schema, before starting the daemon ```sudo service mycodo start```

My testing last night had PWM and PID working fine, as far as I could tell, so let me know of any issues.

**Comment by kizniche** [comment]
I just added better error reporting to all sensors, so if you could update and wait for the error to happen again, there should be a more detailed error in your log to identify the issue.

**Comment by wangine** [comment]
I testing right now on other RPi , with a simple PID and a laptop fan controlled by PWM to cooling the CPU , sensor is internal RPi. I still wait . Lover direction is working?

**Comment by kizniche** [comment]
Lower should be working. I've only been testing the raise option, but lower should theoretically work.

**Comment by wangine** [comment]
Yea , raise work, lower not. Also like a suggestion, should put limit of math between 1-100 . We dont need 21,233333333 duty cycle. Is unuseless stress for RPi.  In C,  i use like example:  for float operations `32.0f * 1.663 = 53.3 not 53.3216 , result is rounded  `
I dont know how that's is doing in python, but you know for sure. 
Also i didn't notice anything wrong below 2 duty . in 1000 samples at 10kHz no any pulse miss or clock fail. 
Still graph and live page i thing was not updated

**Comment by kizniche** [comment]
I'm fairly sure there isn't any benefit to rounding. It may in fact require more computation to do so, considering the input value is a float anyway. Rounding a float would still produce a float as an input. Where did you get the 21,233333333 value?

I just finished getting graphs and the live page to work and pushed a commit.

**Comment by kizniche** [comment]
Graphs look like they're working nicely. Lower output should now work, but I haven't tested it.

![screenshot-192 168 0 5-2017-07-20-15-19-40](https://user-images.githubusercontent.com/838427/28434868-0414821e-6d5f-11e7-8422-724ed66d0134.png)

**Comment by wangine** [comment]
When PID is active if you look on relay stats we can see what value duty cycle have. Yeaa graph look great . 
I can do any test right now because i formated my RPi again with fresh jessie image. Is strange , wasn't succes on clean install `sudo apt-get install -y jq
cd ~
curl -s https://api.github.com/repos/kizniche/Mycodo/releases/latest | \
jq -r '.tarball_url' | wget -i - -O mycodo-latest.tar.gz
mkdir Mycodo
tar xzf mycodo-latest.tar.gz -C Mycodo --strip-components=1
rm -f mycodo-latest.tar.gz
cd Mycodo/install
sudo /bin/bash ./setup.sh`
I dont know what happen , also i try to install tle old orhive of Mycodo just to check if why daemon still stop without reason. I'm still trying.
`Mycodo-master03.10.2016.zip`  That was the best version without any bug, still running on other RPi from that date without any error. I dont know what happen

**Comment by kizniche** [comment]
Yeah, I thought that would be a nice feature to see what duty cycle the pin is outputting on the output page. It's very helpful if you don't want to load the whole graph page. The outputs page is currently set to a 6 second refresh period.

The install instructions in the README will download the latest release, not the latest commit from the PWM_support branch. To install from the PWM_support branch, you'll have to use the instructions in [this comment](https://github.com/kizniche/Mycodo/issues/262#issuecomment-316564431).

**Comment by kizniche** [comment]
I'll be releasing 5.1.0, which includes the PWM support, soon. I just want to make sure there aren't any other bugs lurking. I at least want to test the PWM lower output myself.

**Comment by kizniche** [comment]
...As well as probably add some more PID options for PWM, to allow selecting what duty cycle calculation method to use, and maybe another option to select how the min/max options operate. So far, I think there can be two min/m,ax methods that can be useful:

1. If the PID output is below the min or above the max, set the duty cycle to min or max, respectively.

2. If the PID output is below the min or above the max, set the duty cycle to 0 (turn off PWM).

**Comment by kizniche** [comment]
The latest PWM_support version has a lot of new sensor error reporting. So, if you experience an error with that version, we may be able to find the issue and correct it to stop the daemon form crashing.

**Comment by wangine** [comment]
The first option with gradualy increase should be beter because controll can come almost linear and perfect if PID is tunned right. Also lower dont work for sure because i tested today on other PI with new image and Mycodo , Rise work good. Anyway the time will decide .
Still great work

**Comment by kizniche** [comment]
I fixed the lower PID output 30 minutes ago with commit 37627aae

**Comment by wangine** [comment]
Cant try right now, i still install the old release. I want to be sure any hardware and sersors are OK 100% .That release like i say still running from last year, so i hope will finish without errors

**Comment by kizniche** [comment]
Oooouch! It's a version where it has to compile numpy. My condolences 🙏 

Okay, I understand now. So, one decimal place should be sufficient?

**Comment by wangine** [comment]
No , because actualy to duty cycle to be set at 20kHz with one decimal PWM need to be capable to run at 200kHz. No reason of that . Just rounded is perfect , like 32 or 67 duty . Mycodo was running great with relays (is like a 0 and 100% duty cycle teoreticaly the relay modulation)  , with PWM we increase ability with 1000%. I thing is more than enouth on any purpose. If you want to see the real computation , that's the different thing and can be nice to see.
Yea thanks for support about numpy ./setup.sh: line 82: /home/pi/Mycodo/mycodo/scripts/crontab.sh: Permission denied
I get a error anyway

**Comment by kizniche** [comment]
What is that line? Are you running the setup as root?

**Comment by kizniche** [comment]
Not sure about that error, if it's not the current version of Mycodo, I don't really know how to investigate it, unless you have a commit I can reference to see the state of the code at that time.

**Comment by wangine** [comment]
LOL no , was just a joke , i try to install the old Mycodo version , dont bother with . No , i run the setup as pi user. I just try to see if my board is ok and all my sensors . Is the best debug with knowing working code

**Comment by wangine** [comment]
I canot install the old Mycodo. I dont know why . Still on last release i has some trouble. Twice a new fresh install , impossible to jesie full, . Verry hard on lite and only with PWM update . 
PWM seems working. I canot test right now on real hardware to see how he do the job, but i look in daemon log soon i install and i configure random PWm relay.
`2017-07-21 03:27:42,910 - mycodo.pid_1 - INFO - Activated in 306.6 ms
2017-07-21 03:28:04,460 - mycodo.stats - ERROR - Could not send anonymous usage statistics: {"error":"timeout"}
Traceback (most recent call last):
  File "/var/www/mycodo/mycodo/utils/statistics.py", line 318, in send_anonymous_stats
    client.write_points(formatted_stat_dict)
  File "/var/www/mycodo/env/local/lib/python2.7/site-packages/influxdb/client.py", line 456, in write_points
    tags=tags, protocol=protocol)
  File "/var/www/mycodo/env/local/lib/python2.7/site-packages/influxdb/client.py", line 506, in _write_points
    protocol=protocol
  File "/var/www/mycodo/env/local/lib/python2.7/site-packages/influxdb/client.py", line 302, in write
    headers=headers
  File "/var/www/mycodo/env/local/lib/python2.7/site-packages/influxdb/client.py", line 259, in request
    raise InfluxDBServerError(response.content)
InfluxDBServerError: {"error":"timeout"}

2017-07-21 03:28:13,293 - mycodo.relay - WARNING - PWM duty cycle must be a positive value
2017-07-21 03:28:43,175 - mycodo.relay - WARNING - PWM duty cycle must be a positive value
2017-07-21 03:29:13,264 - mycodo.relay - WARNING - PWM duty cycle must be a positive value`
and error stil going down .
Of course duty can't be negative value. I assume that happen in PID integrator. I dont get it.
Even on lower is the same , possitive value 
raise duty 100 75 50 25 --- setpoint --- 25 50 75 100  lower duty

**Comment by kizniche** [comment]
Try [this](https://github.com/kizniche/Mycodo/issues/275#issuecomment-316867631) amd see if it fixes your influxdb issue.

**Comment by wangine** [comment]
Yea , maybe that was the problem with my sensors and stop daemon , because every time i look in daemon log, InfluxDB became first error and below my sensors stop to read .

**Comment by kizniche** [comment]
It was probably a combination of the sensor controller not being able to write to the influx db and the PID controllers, web UI, etc. not being able to read the measurements in the influx db. Nothing directly takes measurements except for the sensor controller, to ensure measurements aren't interrupted by a process. Everything else feeds from the influxdb.

**Comment by wangine** [comment]
Yes , you assume right . 
`2017-07-16 13:31:51,415 - mycodo.pid_1 - WARNING - No data returned from influxdb
2017-07-16 13:32:22,801 - mycodo.sensors.sht1x_7x - ERROR - SHT1x7xSensor raised an exception when taking a reading: (113L, 255)
2017-07-16 13:36:22,517 - mycodo.sensors.sht1x_7x - ERROR - SHT1x7xSensor raised an exception when taking a reading: (214L, 255)
2017-07-16 14:10:45,435 - mycodo - INFO - Received command to terminate daemon`
Like that error became. 
I will test tomorow on a real hardware the PWM with hard stress for him...

**Comment by kizniche** [comment]
I fixed the lower output with the latest commit and tested both raise and lower work as expected. Here's a graph of the response to me heating the sensor with my finger and then letting go for it to return to ambient temperature.

![screenshot-192 168 0 5-2017-07-20-22-18-24](https://user-images.githubusercontent.com/838427/28446579-ac5efafa-6d99-11e7-9aa6-1e755dc48b22.png)

**Comment by wangine** [comment]
Look realy nice. Unfortunately my daemon still stop . Was a fresh install on jessie lite , i dont know what is wrong. The same error on many lines below.
`Traceback (most recent call last):
  File "/var/www/mycodo/mycodo/sensors/sht2x.py", line 118, in read
  File "/var/www/mycodo/mycodo/sensors/sht2x.py", line 88, in get_measurement
  File "/var/www/mycodo/env/local/lib/python2.7/site-packages/smbus/smbus.py", line 56, in __init__
  File "/var/www/mycodo/env/local/lib/python2.7/site-packages/smbus/smbus.py", line 83, in open
IOError: 24
2017-07-21 13:06:46,090 - mycodo.sensors.raspi - ERROR - RaspberryPiCPUTemp.get_measurement() method raised IOError: [Errno 24] Too many open files: '/sys/class/thermal/thermal_zone0/temp'
2017-07-21 13:06:48,952 - mycodo.sensors.sht2x - ERROR - SHT2xSensor raised an exception when taking a reading: 24`
![graph](https://user-images.githubusercontent.com/20494947/28459799-b62ec1ae-6e17-11e7-8140-41984be935e5.png)
![live stilll work](https://user-images.githubusercontent.com/20494947/28459817-bfab943c-6e17-11e7-8553-7035a99f83be.png)
It is weird because daemon is stopped but sensors still measure on live page. 
When i install Mycodo i get a single error , i canot expand file system. Any clue?

**Comment by kizniche** [comment]
This is a new, rare error, and is a duplicate issue of #254 that I have never experienced myself or been able to replicate.

**Comment by kizniche** [comment]
If you cannot expand the filesystem, then do not continue with the Mycodo install. THere will be problems. You need to be able to expand the filesystem to have enough space to do anything on the SD card or linux.

**Comment by wangine** [comment]
First off all. I try to put Mycodo on fresh jessie install. I notice on first boot of RPi after write image , short message `The system memorry will be resised,,, reboot`
after i get error trying to expand filesystem and install mycodo.
Trying to install old version (that version still running fron last year on other RPi)
![last version](https://user-images.githubusercontent.com/20494947/28484485-4c2d6684-6e7b-11e7-9ae9-424d52077b70.png)
Many errors , also trying to write permissions of a file. So far so good , alot of frustration.
I found a old Jessie full image on my other PC, so i put it on SD card. Strange when i boot that message with resize file system didn't appear. So i go on expand file system and i get succes.
I again to install last version. without succes,`./setup.sh: line 123: /home/pi/Mycodo/mycodo/scripts/update_mycodo.sh: Permission denied
` so i write manualy the file permisions in a tree. After that i get succes install. Running from last night without daemon error or influxDB error. In conclusion my sensors are ok , is the promem in last versions of Mycodo. I notice that error after first update in february or march 5.0.16 (2017-03-30) . I dont remember for sure. From that moment daemon still crash once off 2 or 3 days , but after restart everything become to normal others 2 or 3 days. I dont know where is the mistake. I didn't made a new issue because i didn't think is important but i realise , now after so many versions will be more dificult fo find it.

**Comment by kizniche** [comment]
I don't have an issue with any of my multiple systems of the daemon spontaneously crashing, so I won't be able to directly help in finding the issue, because I cannot replicate it. I can only increase the logging n hopes there is a relevant log line that is produced on your system.

**Comment by wangine** [comment]
Yea. Thanks for that . I have a free RPi now for tests, i need to receive the two sensors SHT75 and 71 to made the real test. If i pull out the SHT25 or 75 in last version of Mycodo , the daemon will crash after some minutes and some time instantly , In those old versions even the sensors are dead daemon will still running. 
![sensor pull out](https://user-images.githubusercontent.com/20494947/28485229-b458cbae-6e80-11e7-9977-87145bc46390.png)
Example here, the sensor are pulled out and system dont crash. 
`I don't have an issue with any of my multiple systems of the daemon spontaneously crashing,`
Did you try to install the last version on a fresh Jessie image?
You will get error on first try , i try it several times on two differents SD cards and two RPi boards in last two days.Still maybe you have right about "expand file system".
I will try in hard way , manualy with fdisk as sudo.

**Comment by wangine** [comment]
No , i think my partitions are OK. 
`pi@raspberrypi:~ $ df
Filesystem     1K-blocks    Used Available Use% Mounted on
/dev/root       29766524 4407668  24089080  16% /
devtmpfs          435668       0    435668   0% /dev
tmpfs             440276       0    440276   0% /dev/shm
tmpfs             440276    6108    434168   2% /run
tmpfs               5120       8      5112   1% /run/lock
tmpfs             440276       0    440276   0% /sys/fs/cgroup
/dev/mmcblk0p1     64456   22240     42216  35% /boot
tmpfs              88056       0     88056   0% /run/user/1000
`
and on second one with last version
`pi@raspberrypi:~ $ df
Filesystem     1K-blocks    Used Available Use% Mounted on
/dev/root       15473580 2061164  12752460  14% /
devtmpfs          468148       0    468148   0% /dev
tmpfs             472756       0    472756   0% /dev/shm
tmpfs             472756   12236    460520   3% /run
tmpfs               5120       8      5112   1% /run/lock
tmpfs             472756       0    472756   0% /sys/fs/cgroup
/dev/mmcblk0p1     42030   21418     20612  51% /boot
`
Files are expanded

**Comment by kizniche** [comment]
> If i pull out the SHT25 or 75 in last version of Mycodo , the daemon will crash after some minutes

This is unexpected. There are ```try```/```except``` statements encapsulating the measurement code, so the daemon shouldn't crash. I'll review the code a little later when I get back to a computer.

**Comment by wangine** [comment]
So, you sugest is not the sensors problem , is the system it selfs. Ok i will give the last try with last version , anyway tomorow i go to made some clean of my sevice and i will be able to test 6 - 7 hours the PWM code in different conditions , AC DC , motors LED's ....... Can wait untill tomorow night and can close that issue . Thanks for your time

**Comment by kizniche** [comment]
To my understanding, I set up the daemon in a way that if any sensor fails, it cannot bring down any other part of the daemon. In fact, each sensor controller is it's own thread. If ti does fail, at worst, that thread (and controller) should just stop working, and not affect anything else. I'll see if there's anything I overlooked when designing those sensor modules. And you have confirmed that unplugging both sensors causes the daemon to crash, or has this not been confirmed as 100% causing a crash?

**Comment by wangine** [comment]
I canot be sure 100% , because if i install a new fresh Mycodo and i will dont activate any SHT sensors , daemon will running without any problem for days. Soon i activate that type of sensors i get the errors and crash. If i pulled out daemon dont stay alive more that 5 minutes.If i let running daemon will stop randomly That i tested in last 3 days , with alot of fresh install. I can confirme 100% thats sensors SHT 25/75 are 100% good because i changed fom other running system with 4.0.10 version of mycodo.

**Comment by wangine** [comment]
I can confirm PWM_support work as expect .I notice , is more dificult to tune the PID but is more faster than relay modulation. The test are made trying to force proccesor to get hot , and  work realy great . 
![output pid pwm working](https://user-images.githubusercontent.com/20494947/28493161-57532446-6f1a-11e7-87da-62d74cb14398.png)
Tunning are not realy easy but PID work more faster.
![pwm pid working](https://user-images.githubusercontent.com/20494947/28493171-7a8a7e96-6f1a-11e7-9fa9-9a05b72e077f.png)
The graph look nice . So, here in internal RPi sensor get fast changes and made the PID to do a very hard and difficult job, but succesful 
![pwm graph confirm working](https://user-images.githubusercontent.com/20494947/28493184-c4b5ad38-6f1a-11e7-9675-e5566be7feb4.png)

![pwm graph confirm 2](https://user-images.githubusercontent.com/20494947/28493187-cbd2268c-6f1a-11e7-92a9-51ce4c434e81.png)

Still trying to add the normal wired relay (not PWM ) dont working. 

![add relay wired](https://user-images.githubusercontent.com/20494947/28493197-fb77aec0-6f1a-11e7-9fba-8ff41e074683.png)
I dont know why , maybe because i swith to PWM_support branch ,?

**Comment by kizniche** [comment]
You found a bug. I'll have to diagnose it later. Thanks.

**Comment by wangine** [comment]
Yes , working now. I keep testing in different conditions, i realy want to be stable.

**Comment by wangine** [comment]
Is strange . I let be tested and when i wake up , Mycodo still running but i can't acces the webserver . Also SSH dont working. I try on differents PC. Seems was orking on PRi web interface itself. I look in logs , no daemon loogs, but some lines on HTTP log.
`[Sun Jul 23 06:25:14.049833 2017] [ssl:warn] [pid 1514:tid 1995751424] AH01909: 127.0.1.1:443:0 server certificate does NOT include an ID which matches the server name
[Sun Jul 23 06:25:14.050737 2017] [wsgi:warn] [pid 1514:tid 1995751424] mod_wsgi: Compiled for Python/2.7.8.
[Sun Jul 23 06:25:14.050779 2017] [wsgi:warn] [pid 1514:tid 1995751424] mod_wsgi: Runtime using Python/2.7.9.
[Sun Jul 23 06:25:14.052281 2017] [mpm_event:notice] [pid 1514:tid 1995751424] AH00489: Apache/2.4.10 (Raspbian) OpenSSL/1.0.1t mod_wsgi/4.3.0 Python/2.7.9 configured -- resuming normal operations
[Sun Jul 23 06:25:14.052380 2017] [core:notice] [pid 1514:tid 1995751424] AH00094: Command line: '/usr/sbin/apache2'`

that was before to restart 

`[Sun Jul 23 14:44:36.832078 2017] [mpm_event:notice] [pid 1514:tid 1995751424] AH00491: caught SIGTERM, shutting down
[Sun Jul 23 14:44:49.578115 2017] [ssl:warn] [pid 838:tid 1996300288] AH01909: 127.0.0.1:443:0 server certificate does NOT include an ID which matches the server name
[Sun Jul 23 14:44:49.655309 2017] [ssl:warn] [pid 842:tid 1996300288] AH01909: 127.0.0.1:443:0 server certificate does NOT include an ID which matches the server name
[Sun Jul 23 14:44:49.708948 2017] [wsgi:warn] [pid 842:tid 1996300288] mod_wsgi: Compiled for Python/2.7.8.
[Sun Jul 23 14:44:49.709042 2017] [wsgi:warn] [pid 842:tid 1996300288] mod_wsgi: Runtime using Python/2.7.9.
[Sun Jul 23 14:44:49.713878 2017] [mpm_event:notice] [pid 842:tid 1996300288] AH00489: Apache/2.4.10 (Raspbian) OpenSSL/1.0.1t mod_wsgi/4.3.0 Python/2.7.9 configured -- resuming normal operations
[Sun Jul 23 14:44:49.714010 2017] [core:notice] [pid 842:tid 1996300288] AH00094: Command line: '/usr/sbin/apache2'` 
And that after i restart.
Also i notice a error on influxDB , but seems are ok , just a delay start (i think);
`2017-07-23 14:45:37,140 - mycodo - INFO - Mycodo daemon v5.0.49 starting
2017-07-23 14:45:37,431 - mycodo - INFO - Anonymous statistics disabled
2017-07-23 14:45:37,698 - mycodo.output - INFO - PWM 1 setup on pin 16
2017-07-23 14:45:37,699 - mycodo.output - INFO - Relay 2 setup on pin 12 and turned OFF (OFF=LOW)
2017-07-23 14:45:37,701 - mycodo.output - INFO - Output controller activated in 100.1 ms
2017-07-23 14:45:37,701 - mycodo - INFO - All activated timer controllers started
2017-07-23 14:45:37,905 - mycodo.sensor_1 - INFO - Activated in 158.9 ms
2017-07-23 14:45:37,907 - mycodo - INFO - All activated sensor controllers started
2017-07-23 14:45:38,179 - mycodo.pid_1 - INFO - Activated in 227.1 ms
2017-07-23 14:45:38,282 - mycodo - INFO - All activated PID controllers started
2017-07-23 14:45:38,283 - mycodo - INFO - All activated LCD controllers started
2017-07-23 14:45:38,284 - mycodo - INFO - Mycodo daemon v5.0.49 started in 1.144 seconds
2017-07-23 14:45:38,296 - mycodo - INFO - 41.888 MB ram in use
2017-07-23 14:46:02,321 - mycodo.influxdb - ERROR - No measurement available in the past 60 seconds.
2017-07-23 14:46:02,321 - mycodo.pid_1 - WARNING - No data returned from influxdb`
The influxDB was downgraded , as you suggest.

**Comment by kizniche** [comment]
I just noticed that influxdb 1.3.1 was released 3 days ago (without updating their changelog for me to notice). I just upgraded and it appears to be working. I used the following to upgrade:

```bash
wget https://dl.influxdata.com/influxdb/releases/influxdb_1.3.1_armhf.deb
sudo dpkg -i influxdb_1.3.1_armhf.deb
sudo service influxdb restart
rm -rf influxdb_1.3.1_armhf.deb
```

**Comment by hamster007** [comment]
Hello, 
This is an awesome feature, Just a quick question. 
Will this control a extraction fan PWM with little to no adjustment?
Thanks in advance.

**Comment by kizniche** [comment]
I'm not sure what that is. Currently, only PIDs can output a PWM signal. If your device can accept a 3.5-volt PWM from 0% to 100% duty cycle, it will work out of the box. Otherwise you'll need to adjust your hardware accordingly.

**Comment by hamster007** [comment]
Its a standard 12v 4wire computer fan. Wiring: 12v, ground, tachometer, PWM.
The Idea is to be able to utilise the PWM output from mycodo to control temperatures without turning relays on/off every few seconds. This will give better temp control as well as lowering power draw due to constant running state or off if below temp theshold

**Comment by zsole2** [comment]
I have followed this development too, for the exact same purpose, i.e. for controlling fans.  
The problem is that there is a minimum voltage  that is needed for the fan to turn on, in my experience it is about 4-5 V for a 12 V fan, thereby affecting the duty cycle range. This would only require a small modification to the code, though... I still have a DIY picture frame running based on the setup described at http://www.raspberrypi.org/forums/viewtopic.php?f=35&t=62438, if that is helpful.

**Comment by Magnum-Pl** [comment]
I have a few 4-wire PWM Delta server fans that accept a minimum control signal voltage of 3.3v. Check your datasheet to see what specs your fan has. I'll test one of mine this weekend and post the results. I'm currently using a cheap DC PWM motor controller to control my fans, but using Mycodo would be a lot better.

**Comment by pukkita** [comment]
> Ok. The first complete schematics with bipolar transistors, to controll with PWM and isolated output .
> The normal circuit in phase with PWM , the most stable with bipolar transistor
> pwm 50 duty bipolar isolated no inverting phase
> And inverted phase than PWM ,
> pwm 50 duty bipolar isolated 1
> Are tested only on resistive loads . Not on motors (fully inductive)

Will this work with fans? Would be great to be able to PWM drive an inline AC fan (RVK125A) for extraction!

**Comment by kizniche** [comment]
v5.3.4 released today includes the ability to create timers that output a PWM signal based on a method. Just make sure your method stays between 0 and 100 (%), and select it in a PWM Timer to use it.

**Comment by pukkita** [comment]
Regarding PWM commanding AC inductive motors (Fans), according to [RVK](https://planetaklimata.com.ua/instr/Systemair/Systemair_Fans_RVK_Data_sheet_Eng.pdf) it can be done using stepless thyristors:

> The fans can be speed-controlled via a stepless thyristor or a 5-step transformer.

**Comment by Theoi-Meteoroi** [comment]
Lets just say - controlling large motors is out of scope I think.  Indeed, many things _can_ be done with proper signal conditioning and isolation from the big, bad motor.  The problem is that as you try to scale the solution (engineer to reality) you figure out you started from the wrong perspective.  Tiny 3.3v digital signals need optoisolators at a minimum and the power stage is a challenge to get done well.  At that point you couldn't care less about what wiggles the control signal.  I would try to find an engineered solution that takes a 3.3v control signal first then consider building my own if that option isn't feasable if I want to integrate with Mycodo.  There is only so much of the effort that Kyle can sustain, I think.

**Comment by victorhooi** [comment]
Not sure if I should post here, or start a new issue.

However, the original comment talked about dimming some Meanwell LED drivers, which required 0-10V for the PWM.

I have the same use-case (Meanwell LED dimming using PWM) - Raspberry Pi's do 0-5V for PWM, right?

What are you all using to convert that to 0-10V PWM?

The post did mention some converter, however, the [link](https://www.uctronics.com/0-5v-0-10v-to-0-100-voltage-to-pwm-converter-module.html) there is dead.

**Comment by kizniche** [comment]
There have been a few discussions about generating 10 VDC PWM signals, but these same hardware solutions can also do 5 VDC.

https://forum.radicaldiy.com/t/schematic-using-pwm-to-generate-a-0-10v-control-signal/1098

https://forum.radicaldiy.com/t/can-i-control-the-speed-of-motors-and-fans-with-a-pwm-signal/65

**Comment by victorhooi** [comment]
@leoneo Thanks for the link - but just to clarify - that is a digital potentiometer board, right? Not for converting a 5V PWM to 10V PWM? Is the idea that we use this for one of the other non-PWM Meanwell dimming modes?

@kizniche Cool - thank you for that as well. The second link, which then links to a [OSHPark PCB](https://oshpark.com/shared_projects/oJIFTsvc) seems like exactly what I want.

Sorry if these are basic questions -  but my electronics knowledge is fairly basic.

But to confirm, I can just:
1. Order this board from OSHPARK
2. Solder on the resisters and 2 x transistors
3. Connect the PWM and GND on left to the Raspberry Pi (**I'm not too clear on the connectivity here**)
4. Connect the two pins on the right to the two dimming pins on a Meanwell driver.

And then it should just work as-is to dim Meanwell drivers via PWM, right? Do you know if anybody has tried or confirmed this circuit with Meanwell drivers?

**Comment by victorhooi** [comment]
@kizniche I've just ordered a few of your board from OSHPark =).

Sorry to bother you - but just to confirm - this should just work with the Meanwell 0-10V dimmers, right?

Is there anything special I need to adapt your board for this use-case?

**Comment by kizniche** [comment]
>Connect the PWM and GND on left to the Raspberry Pi

Correct

>Connect the two pins on the right to the two dimming pins on a Meanwell driver.

On the right side of the board, you will connect 5 VDC (instead of 10 VDC), ground, and the 5 VDC PWM signal will be generated on the PWM pin.

Devices are either controlled by a 0 - some voltage OR a PWM signal, not both. This will generate a PWM signal at whater input voltage you use, either 5 VDC or 10 VDC.

**Comment by victorhooi** [comment]
I just received the PCB's today =) - yay!

![PXL_20221005_115547111](https://user-images.githubusercontent.com/243499/194060311-7eed127f-0495-454e-9a80-8b0ae3f74ae0.jpeg)

As far as I can tell, the Meanwell PWM LED dimming mode needs a **10V** PWM signal:
* [Spec sheet](https://www.meanwell.com/Upload/PDF/HLG-100/HLG-100-SPEC.PDF) (page 5 - 10V PWM, 100Hz to 3Khz
* [Discussion](https://electronics.stackexchange.com/questions/461049/meanwell-3-in-1-dimming)).

The one part I still don't get is the right-hand side connectivity 😕.

@kizniche You mention an input voltage - does this mean I need to supply a 10V voltage to the board, on the right hand side connectors somewhere, for it to convert the Raspberry Pi 5V PWM to a 10V PWM signal?

Ugh, if so, that could be a problem. Where do people normally get this, if they just use a Raspberry Pi?

**Comment by kizniche** [comment]
If using a Pi, you're not converting from 5 VDC. GPIO pins output 3.3 VDC. 

Usually the device can provide the higher DC voltage. For instance, my Terrabloom and AC Infinity fans that can be speed controlled via 10 VDC PWM have 10 VDC as one of the pins of the control wire (GND, 10 VDC, PWM signal in).

You could use the 5 VDC of the Pi if you don't want to use an external power source.

**Comment by victorhooi** [comment]
Thanks so much for your help so far =).

You're right, Raspberry Pi PWM output is 3.3V, not 5V (I think Arduino is 5V?)

I've checked - the Meanwell LED dimmers don't appear to supply 10VDC anywhere for me to tap into.

Ah - so you're thinking of taking 5VDC from pin 2 or 4 on the Raspberry Pi, right? I'd need to connect 5V and GND from the Raspberry Pi to the 10V and GND pins on the right-side of the board, correct?

But that means my resulting PWM output will be 0-5V, right? There's no way to get that to 10V?

**Comment by kizniche** [comment]
>I'd need to connect 5V and GND from the Raspberry Pi to the 10V and GND pins on the right-side of the board, correct?

GND is common across all GND points on the board, so it doesn't matter where you connect it.

>But that means my resulting PWM output will be 0-5V, right? There's no way to get that to 10V?

like I mentioned previously, you can either produce a varying voltage signal, e.g. 0 - 10 volts, or a PWM signal. You cannot do both. PWM signals are one single, static voltage. There are no drivers that I know of that require a varied voltage PWM signal, it's not just a thing.

**Comment by victorhooi** [comment]
I did do some research before, and came across one thread which seems to offer a solution  using just the Raspberry Pi and no separate supply - I confess I don't quite understand all of the details yet:

* [Forum thread](https://forums.raspberrypi.com/viewtopic.php?t=294921)
* [Final Circuit Diagram](https://www.circuit-diagram.org/circuits/7d4ed4ab1c4449ba98752068f095c167)
* [Corresponding Reddit thread](https://www.reddit.com/r/AskElectronics/comments/kd96xb/raspberry_pi_gpio_pwm_to_dim_mean_well_led_driver/)

which references [this earlier Reddit thread](https://www.reddit.com/r/led/comments/966f0c/comment/e402coy/):

> The challenge in your case is figuring out how to make the Raspberry Pi 5V output signal talk to the Meanwell which is expecting 10V. When using the 10V PWM method, the Meanwell actually provides its own 10V signal (via the blue wire labelled DIM+), so the good news is you don't need an external 10V power source.

> What you can do is use an NPN transistor as a switch to convert the Raspberry Pi 5V PWM signal into a 10V PWM signal. Since PWM is, in essence, just a rapid switching between on and off states, the transistor approach will allow you to "convert" the 5V PWM to a 10V PWM signal. Whatever duty cycle is set by your Pi at 5V will be used as an input signal into the Meanwell HLG at 10V, which will then interpret the signal for dimming as specified in the datasheet.

So there's a 10V differential across the Meanwell DIM+ and DIM- pins.

Is there some way to connect that to your circuit board? =)
